### PR TITLE
Feature/gridedit 934 undo for curvilinear

### PR DESCRIPTION
**Pull request by BillSenior**

### CODE REVIEW

**Review by lucacarniato — APPROVED** [review]

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp [hunk 1]
```diff
@@ -232,23 +340,34 @@ namespace meshkernel
...
-        void MoveNode(Point const& fromPoint, Point const& toPoint);
+        [[nodiscard]] UndoActionPtr MoveNode(Point const& fromPoint, Point const& toPoint);
 
         /// @brief Get the mesh bounding box.
         BoundingBox GetBoundingBox() const;
 
         /// @brief The number of nodes M in the m dimension
         /// @return A number >= 2 for a valid curvilinear grid
-        UInt NumM() const { return static_cast<UInt>(m_gridNodes.cols()); }
+        UInt NumM() const { return static_cast<UInt>(m_gridNodes.cols()) - m_startOffset.m_m - m_endOffset.m_m; }
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp:360]: The first option is correct

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp [hunk 2]
```diff
@@ -284,12 +444,30 @@ namespace meshkernel
...
-                     CurvilinearGridNodeIndices const& secondNode);
-
-        lin_alg::Matrix<Point> m_gridNodes;                    ///< Member variable storing the grid
-        lin_alg::Matrix<bool> m_gridFacesMask;                 ///< The mask of the grid faces (true/false)
-        lin_alg::Matrix<NodeType> m_gridNodesTypes;            ///< The grid node types
+        [[nodiscard]] UndoActionPtr AddEdge(CurvilinearGridNodeIndices const& firstNode,
+                                            CurvilinearGridNodeIndices const& secondNode);
+
+        lin_alg::Matrix<Point> m_gridNodes; ///< Member variable storing the grid
+        char unusedArray1[2049];
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp:451]: all debugging members and output has been removed

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/ResetCurvilinearNodeAction.hpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        CurvilinearGridNodeIndices m_nodeId;
+
+        /// @brief The initial node
+        Point m_initialNode;
+
+        /// @brief The updated node
+        Point m_updatedNode;
+
+        /// @brief Indicate if the node types need to be recomputed after undo/redo
+        bool m_recalculateNodeTypes;
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/ResetCurvilinearNodeAction.hpp:85]: Removed

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp [hunk 4]
```diff
@@ -175,11 +242,49 @@ namespace meshkernel
...
+            return m_gridFacesMask(n + m_startOffset.m_n, m + m_startOffset.m_m);
+        }
+
+        void SetIsFaceMaskValid(UInt n, UInt m, bool value)
+        {
+            m_gridFacesMask2.at(n).at(m) = value;
+            // m_gridFacesMask(n + m_startOffset.m_n, m + m_startOffset.m_m) = value;
+        }
+
+        // bool& IsFaceMaskValid(UInt n, UInt m)
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp:268]: Now uncommented, was for debugging purposes

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridRefinementUndoAction.cpp [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    : BaseMeshUndoAction<CurvilinearGridRefinementUndoAction, CurvilinearGrid>(grid), m_nodes(grid.GetNodes()), m_lower(grid.StartOffset()), m_upper(grid.EndOffset()) {}
+
+void meshkernel::CurvilinearGridRefinementUndoAction::Swap(lin_alg::Matrix<Point>& nodes, CurvilinearGridNodeIndices& lower, CurvilinearGridNodeIndices& upper)
+{
+    lin_alg::Matrix<Point> temp = nodes;
+    nodes = m_nodes;
+    m_nodes = temp;
+    std::swap(m_lower, lower);
+    std::swap(m_upper, upper);
+    // m_nodes.swap(nodes);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridRefinementUndoAction.cpp:19]: Reverted code

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp [hunk 6]
```diff
@@ -44,28 +50,28 @@ namespace meshkernel
...
-            UpperRight,    //(13)
-            Left,          //(4)
-            Right,         //(2)
-            Bottom,        //(1)
-            Up,            //(3)
-            InternalValid, //(10)
-            Invalid        //(0)
-        };
+        // /// @brief An enum for curvilinear node types
+        // enum class NodeType
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp:54]: REmoved

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp [hunk 7]
```diff
@@ -222,7 +327,10 @@ namespace meshkernel
         /// @param firstNode The first node of the boundary grid line.
         /// @param secondNode The second node of the boundary grid line.
         /// @return If a new grid line has been allocated
-        bool AddGridLineAtBoundary(CurvilinearGridNodeIndices const& firstNode, CurvilinearGridNodeIndices const& secondNode);
+        std::tuple<bool, UndoActionPtr> AddGridLineAtBoundary(CurvilinearGridNodeIndices const& firstNode, CurvilinearGridNodeIndices const& secondNode);
+
+        void DeleteGridLineAtBoundary(const CurvilinearGridNodeIndices& firstNode,
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp:332]: Removed debugging function

File: libs/MeshKernel/src/CurvilinearGrid/ResetCurvilinearNodeAction.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+{
+    return m_updatedNode;
+}
+
+bool meshkernel::ResetCurvilinearNodeAction::RecalculateNodeTypes() const
+{
+    return m_recalculateNodeTypes;
+}
+
+void meshkernel::ResetCurvilinearNodeAction::Print(std::ostream& out) const
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/ResetCurvilinearNodeAction.cpp:43]: All of them, gone!

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/AddGridLineUndoAction.hpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+                              const CurvilinearGridNodeIndices& endOffset);
+
+        /// @brief Get the start offset
+        const CurvilinearGridNodeIndices& StartOffset() const;
+
+        /// @brief Get the end offset
+        const CurvilinearGridNodeIndices& EndOffset() const;
+
+        /// @brief Print the add grid line action to the stream
+        void Print(std::ostream& out = std::cout) const override;
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/AddGridLineUndoAction.hpp:63]: All of them, gone!

**Review by lucacarniato — APPROVED** [review]

**Pull request merged**